File: Zamba/ZambaWeb2013/zamba2010/Zamba.Web2014/Views/UC/Task/TaskDetail.ascx.vb
Imports Zamba.Core
Imports Zamba.Services
Imports System.Collections.Generic
Imports Zamba.Core.Enumerators

Partial Public Class TaskDetail
    Inherits System.Web.UI.UserControl

    Private _ITaskResult As Zamba.Core.ITaskResult
    Private TaskId As Integer
    Private _userid As Long
    Private _user As Zamba.Core.IUser

#Region "Properties"

    Public Property TaskResult() As ITaskResult
        Get
            Return _ITaskResult
        End Get
        Set(ByVal value As ITaskResult)
            _ITaskResult = value
        End Set
    End Property

    ''' <summary>
    ''' Evento que ejecuta reglas
    ''' </summary>
    ''' <param name="ruleId">ID de la regla a ejecutar</param>
    ''' <param name="results">Tareas a ejecutar</param>
    ''' <remarks></remarks>
    Public Event ExecuteRule(ByVal ruleId As Int64, ByVal results As List(Of Zamba.Core.ITaskResult))

    Private Sub ThrowExecuteRule(ByVal ruleId As Int64, ByVal results As List(Of Zamba.Core.ITaskResult))
        RaiseEvent ExecuteRule(ruleId, results)
    End Sub
#End Region

    Protected Sub Page_Load(ByVal sender As Object, ByVal e As System.EventArgs) Handles Me.Load
        If Page.Session("UserId") Is Nothing = False AndAlso Not (TaskResult Is Nothing) Then
            Page.Title = TaskResult.Name

            _userid = Long.Parse(Session("UserId"))
            _user = Session("User")



            TaskId = CInt(Request("TaskId"))
            Me.HiddenTaskId.Value = TaskId

            Me.HiddenDocID.Value = TaskResult.ID

            If (IsPostBack = False) Then
                ExecuteOpenRules()
            Else
                If String.IsNullOrEmpty(Me.HiddenCurrentFormID.Value) = False Then
                    TaskResult.CurrentFormID = Int64.Parse(Me.HiddenCurrentFormID.Value)
                End If
            End If

            If TaskResult.CurrentFormID > 0 OrElse HasForms(TaskResult.DocTypeId) Then
                Me.HiddenCurrentFormID.Value = TaskResult.CurrentFormID

                Dim viewer As Views_UC_Viewers_FormBrowser
                viewer = LoadControl("../Viewers/FormBrowser.ascx")
                viewer.IsShowing = True
                viewer.TaskResult = TaskResult

                RemoveHandler viewer.ExecuteRule, AddressOf ThrowExecuteRule
                AddHandler viewer.ExecuteRule, AddressOf ThrowExecuteRule

                pnlViewer.Controls.Add(viewer)
            Else
                Dim viewer As Views_UC_Viewers_DocViewer

                viewer = LoadControl("../Viewers/DocViewer.ascx")
                viewer.Result = TaskResult
                pnlViewer.Controls.Add(viewer)
            End If

            Dim ZOptBusines As Zamba.Services.SZOptBusiness = New Zamba.Services.SZOptBusiness()
            Page.Title = DirectCast(ZOptBusines.GetValue("WebViewTitle"), String) + " - " + TaskResult.Name
        End If
    End Sub

    Public Function HasForms(ByVal docTypeId As Int64) As Boolean
        Dim sForms As New SForms()
        Dim result As Boolean = sForms.HasForms(docTypeId)
        sForms = Nothing

        Return result
    End Function

    ''' <summary>
    ''' Ejecuta las reglas de entrada
    ''' </summary>
    ''' <remarks></remarks>
    Private Sub ExecuteOpenRules()
        Try
            Dim List As New List(Of ITaskResult)()
            Dim formId As Int64 = TaskResult.CurrentFormID
            TaskResult.CurrentFormID = 0
            List.Add(TaskResult)

            Try
                Dim sRules As New SRules()

                Dim rules As List(Of IWFRuleParent) = sRules.GetCompleteHashTableRulesByStep(TaskResult.StepId)

                If Not IsNothing(rules) Then
                    For Each rule As IWFRuleParent In rules
                        If (rule.RuleType = TypesofRules.AbrirDocumento) Then
                            ThrowExecuteRule(rule.ID, List)
                        End If
                    Next
                End If

            Catch ex As Exception
                Zamba.AppBlock.ZException.Log(ex)
            End Try

            'Verifica si una DoShowForm no completo el formulario a mostrar
            If (TaskResult.CurrentFormID = 0) Then
                TaskResult.CurrentFormID = formId
            End If
        Catch ex As Exception
            Zamba.AppBlock.ZException.Log(ex)
        End Try
    End Sub
End Class

File: Zamba/ZambaWeb2013/zamba2010/Zamba.Web2014/Views/UC/Task/TaskDetail.ascx.cs

using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using Zamba.Core;
using Zamba.Services;
using Zamba.Core.Enumerators;
using System.Web.UI;
using System.Web.UI.WebControls;
public partial class TaskDetail : System.Web.UI.UserControl
{

    private Zamba.Core.ITaskResult _ITaskResult;
    private int TaskId;
    private long _userid;

    private Zamba.Core.IUser _user;
    #region "Properties"

    public ITaskResult TaskResult
    {
        get { return _ITaskResult; }
        set { _ITaskResult = value; }
    }

    /// <summary>
    /// Evento que ejecuta reglas
    /// </summary>
    /// <param name="ruleId">ID de la regla a ejecutar</param>
    /// <param name="results">Tareas a ejecutar</param>
    /// <remarks></remarks>
    public event ExecuteRuleEventHandler ExecuteRule;
    public delegate void ExecuteRuleEventHandler(Int64 ruleId, List<Zamba.Core.ITaskResult> results);

    private void ThrowExecuteRule(Int64 ruleId, List<Zamba.Core.ITaskResult> results)
    {
        if (ExecuteRule != null)
        {
            ExecuteRule(ruleId, results);
        }
    }
    #endregion

    protected void Page_Load(object sender, System.EventArgs e)
    {
        if (Page.Session["UserId"] != null && TaskResult != null)
        {

            Page.Title = TaskResult.Name;

            _userid = long.Parse(Session["UserId"].ToString());
            _user = (IUser)Session["User"];

            TaskId = Convert.ToInt32(Request["TaskId"]);
            this.HiddenTaskId.Value = TaskId.ToString();

            this.HiddenDocID.Value = TaskResult.ID.ToString();

            if (!IsPostBack)
            {
                ExecuteOpenRules();
            }
            else
            {
                if (string.IsNullOrEmpty(this.HiddenCurrentFormID.Value) == false)
                {
                    TaskResult.CurrentFormID = Int64.Parse(this.HiddenCurrentFormID.Value);
                }
            }

            if (TaskResult.CurrentFormID > 0 || HasForms(TaskResult.DocTypeId))
            {
                this.HiddenCurrentFormID.Value = TaskResult.CurrentFormID.ToString();

                Views_UC_Viewers_FormBrowser viewer = default(Views_UC_Viewers_FormBrowser);
                viewer = (Views_UC_Viewers_FormBrowser)LoadControl("../Viewers/FormBrowser.ascx");
                viewer.IsShowing = true;
                viewer.TaskResult = TaskResult;

                viewer.ExecuteRule -= ThrowExecuteRule;
                viewer.ExecuteRule += ThrowExecuteRule;

                pnlViewer.Controls.Add(viewer);
            }
            else {
                Views_UC_Viewers_DocViewer viewer = default(Views_UC_Viewers_DocViewer);

                viewer = (Views_UC_Viewers_DocViewer)LoadControl("../Viewers/DocViewer.ascx");
                viewer.Result = TaskResult;
                pnlViewer.Controls.Add(viewer);
            }

            Zamba.Services.SZOptBusiness ZOptBusines = new Zamba.Services.SZOptBusiness();
            Page.Title = (string)ZOptBusines.GetValue("WebViewTitle") + " - " + TaskResult.Name;
        }
    }

    public bool HasForms(Int64 docTypeId)
    {
        SForms sForms = new SForms();
        bool result = sForms.HasForms(docTypeId);
        sForms = null;

        return result;
    }

    /// <summary>
    /// Ejecuta las reglas de entrada
    /// </summary>
    /// <remarks></remarks>
    private void ExecuteOpenRules()
    {
        try
        {
            List<ITaskResult> List = new List<ITaskResult>();
            Int64 formId = TaskResult.CurrentFormID;
            TaskResult.CurrentFormID = 0;
            List.Add(TaskResult);

            try
            {
                SRules sRules = new SRules();

                List<IWFRuleParent> rules = sRules.GetCompleteHashTableRulesByStep(TaskResult.StepId);

                if ((rules != null))
                {
                    foreach (IWFRuleParent rule in rules)
                    {
                        if ((rule.RuleType == TypesofRules.AbrirDocumento))
                        {
                            ThrowExecuteRule(rule.ID, List);
                        }
                    }
                }

            }
            catch (Exception ex)
            {
                Zamba.AppBlock.ZException.Log(ex);
            }

            //Verifica si una DoShowForm no completo el formulario a mostrar
            if ((TaskResult.CurrentFormID == 0))
            {
                TaskResult.CurrentFormID = formId;
            }
        }
        catch (Exception ex)
        {
            Zamba.AppBlock.ZException.Log(ex);
        }
    }
    public TaskDetail()
    {
        Load += Page_Load;
    }
}
